File: backend/app/core/date_utils.py
# backend/app/core/date_utils.py
"""
Единственное место, где считаются календарные даты задач.

Правило:
- duration = календарные дни
- dur=1 -> старт и финиш в один день
- следующая зависимая задача стартует на следующий календарный день

Параметр holidays сохранён в сигнатурах для совместимости вызовов, но не используется.
"""
from datetime import date, timedelta


def add_working_days(
    start: date,
    working_days: int,
    holidays: set[date] | None = None,
) -> date:
    """
    Историческое имя: фактически прибавляет календарные дни к дате.
    start=2026-03-13, working_days=1 -> 2026-03-14
    """
    return start + timedelta(days=max(0, working_days))


def working_days_between(
    start: date,
    end: date,
    holidays: set[date] | None = None,
) -> int:
    """
    Историческое имя: фактически считает календарные дни между двумя датами
    (не включая start, включая end).
    """
    return max(0, (end - start).days)


def task_end_date(start: date, working_days: int, holidays: set[date] | None = None) -> date:
    """Дата окончания задачи (включительно)."""
    return start + timedelta(days=max(0, working_days - 1))


def next_task_start_date(start: date, working_days: int, holidays: set[date] | None = None) -> date:
    """Дата старта следующей зависимой задачи после текущей."""
    return add_working_days(start, max(1, working_days), holidays)

File: project-scripts-with-enir (1)/frontend/lib/dateUtils.ts
// frontend/lib/dateUtils.ts
// Зеркало Python date_utils.py — единая логика рабочих дней

export function addWorkingDays(
  start: Date,
  workingDays: number,
  holidays: Set<string> = new Set(),
): Date {
  if (workingDays <= 0) return new Date(start);
  let current = new Date(start);
  let added = 0;
  while (added < workingDays) {
    current.setDate(current.getDate() + 1);
    const iso = current.toISOString().split("T")[0];
    if (current.getDay() !== 0 && current.getDay() !== 6 && !holidays.has(iso)) added++;
  }
  return current;
}

export function taskEndDate(startDate: string, workingDays: number): string {
  return addWorkingDays(new Date(startDate), workingDays).toISOString().split("T")[0];
}

export function workingDaysBetween(start: Date, end: Date): number {
  let count = 0;
  const cur = new Date(start);
  while (cur < end) {
    cur.setDate(cur.getDate() + 1);
    if (cur.getDay() !== 0 && cur.getDay() !== 6) count++;
  }
  return count;
}

export function fmtDate(iso: string): string {
  if (!iso) return "—";
  const d = new Date(iso);
  return `${String(d.getDate()).padStart(2,"0")}.${String(d.getMonth()+1).padStart(2,"0")}`;
}

export function fmtDateTime(iso: string): string {
  if (!iso) return "—";
  const d = new Date(iso);
  return `${fmtDate(iso)} ${String(d.getHours()).padStart(2,"0")}:${String(d.getMinutes()).padStart(2,"0")}`;
}

export function fmtMoney(n: number): string {
  return n?.toLocaleString("ru-RU", { maximumFractionDigits: 0 }) ?? "—";
}
